File: src/bcihand/utils/latency.py
"""Latency instrumentation for the pipeline stages (project brief section 19).

Usage: call `.mark(stage_name)` at each pipeline stage boundary with a
monotonic timestamp; `LatencyProfiler.summary()` reports per-stage and
total signal-to-command latency statistics.
"""
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field

import numpy as np

STAGES = (
    "acquisition",
    "filtering",
    "candidate_detection",
    "classification",
    "double_blink_confirmation",
    "pc_to_esp32_comm",
)


@dataclass
class LatencyProfiler:
    _durations_s: dict = field(default_factory=lambda: defaultdict(list))

    def record(self, stage: str, duration_s: float) -> None:
        self._durations_s[stage].append(duration_s)

    def record_total(self, duration_s: float) -> None:
        self._durations_s["total_signal_to_command"].append(duration_s)

    def summary(self) -> dict:
        out = {}
        for stage, values in self._durations_s.items():
            if not values:
                continue
            arr = np.asarray(values, dtype=np.float64) * 1000.0  # -> ms
            out[stage] = {
                "n": len(arr),
                "mean_ms": float(np.mean(arr)),
                "median_ms": float(np.median(arr)),
                "p95_ms": float(np.percentile(arr, 95)),
                "max_ms": float(np.max(arr)),
            }
        return out

    def report_text(self) -> str:
        lines = ["Latency profile (ms):"]
        for stage, stats in self.summary().items():
            lines.append(
                f"  {stage:28s} n={stats['n']:5d}  mean={stats['mean_ms']:7.3f}  "
                f"median={stats['median_ms']:7.3f}  p95={stats['p95_ms']:7.3f}  max={stats['max_ms']:7.3f}"
            )
        return "\n".join(lines)

File: web/src/core/utils/latency.ts
/**
 * Latency instrumentation — port of utils/latency.py, used by the
 * Diagnostics page (project brief section 24) to report worker/filter/
 * candidate-detector/classifier/total blink-to-command latency.
 */

export interface LatencyStats {
  n: number;
  meanMs: number;
  medianMs: number;
  p95Ms: number;
  maxMs: number;
}

export class LatencyProfiler {
  private readonly durationsS = new Map<string, number[]>();

  record(stage: string, durationS: number): void {
    const arr = this.durationsS.get(stage) ?? [];
    arr.push(durationS);
    if (arr.length > 2000) arr.shift(); // bound memory for long-running sessions
    this.durationsS.set(stage, arr);
  }

  recordTotal(durationS: number): void {
    this.record("total_signal_to_command", durationS);
  }

  summary(): Record<string, LatencyStats> {
    const out: Record<string, LatencyStats> = {};
    for (const [stage, values] of this.durationsS) {
      if (values.length === 0) continue;
      const ms = values.map((v) => v * 1000);
      const sorted = [...ms].sort((a, b) => a - b);
      const mean = ms.reduce((a, b) => a + b, 0) / ms.length;
      const median = sorted[Math.floor(sorted.length / 2)];
      const p95 = sorted[Math.min(Math.floor(sorted.length * 0.95), sorted.length - 1)];
      const max = sorted[sorted.length - 1];
      out[stage] = { n: values.length, meanMs: mean, medianMs: median, p95Ms: p95, maxMs: max };
    }
    return out;
  }
}
